File: 01jupyter/nb/source/14_montecarlo/problems/00_gacha/gacha.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>

/* 試行ごとに独立な乱数生成器 (MINSTD 法)。
   16807*s は 64bit に収まるので桁あふれしない。返り値は [0,N) の整数。 */
static inline int draw(long long * s, int N) {
  *s = (16807LL * *s) % 2147483647LL;
  return (int)(*s % N);
}

/* 1試行: N 種類の景品が等確率で出るとき, 全種類そろうまでに引いた回数。
   collected を 64bit のビットマスクで管理する (N <= 62 を想定)。 */
long one_trial(int N, long long seed) {
  long long s = seed % 2147483647LL;
  if (s <= 0) s += 2147483646LL;
  unsigned long long got = 0;
  unsigned long long full = (N == 64 ? ~0ULL : ((1ULL << N) - 1));
  long draws = 0;
  while (got != full) {
    int k = draw(&s, N);
    got |= (1ULL << k);
    draws++;
  }
  return draws;
}

int main(int argc, char ** argv) {
  int  N = (argc > 1 ? atoi(argv[1]) : 10);          /* 景品の種類数 */
  long T = (argc > 2 ? atol(argv[2]) : 1000000);     /* 試行回数 */
  double sum = 0.0, sumsq = 0.0;

  /* T 回の試行は互いに独立。各試行の引き回数 d を集計する。 */
  // BEGIN ANSWER: 各試行は独立なので #pragma omp parallel for reduction(+:sum,sumsq) で並列化・集計せよ.
#pragma omp parallel for reduction(+:sum,sumsq)
  // END ANSWER
  for (long t = 0; t < T; t++) {
    long d = one_trial(N, (long long)(t + 1) * 2654435761LL + 12345LL);
    sum   += (double)d;
    sumsq += (double)d * (double)d;
  }

  double mean = sum / T;
  double var  = sumsq / T - mean * mean;
  /* 理論期待値 = N * H_N (H_N は調和数) */
  double H = 0.0;
  for (int k = 1; k <= N; k++) H += 1.0 / k;
  printf("N=%d, trials=%ld: 平均 %.3f 回 (理論 %.3f), 標準偏差 %.3f\n",
         N, T, mean, N * H, sqrt(var));
  return 0;
}

File: 01jupyter/nb/source/14_montecarlo/problems/00_gacha/gacha.f90
module gacha_mod
contains
  ! 試行ごとに独立な乱数生成器 (MINSTD 法). 16807*s は 64bit に収まる.
  ! s を更新し, [0,N) の整数を返す.
  function draw(s, N) result(k)
    integer(8), intent(inout) :: s
    integer, intent(in) :: N
    integer :: k
    s = modulo(16807_8 * s, 2147483647_8)
    k = int(modulo(s, int(N, 8)))
  end function draw

  ! 1試行: N 種類が等確率で出るとき, 全種類そろうまでに引いた回数.
  ! collected を 64bit 整数のビットマスクで管理 (N <= 62 を想定).
  function one_trial(N, seed) result(draws)
    integer, intent(in) :: N
    integer(8), intent(in) :: seed
    integer(8) :: draws, s, got, full
    integer :: k
    s = modulo(seed, 2147483647_8)
    if (s <= 0) s = s + 2147483646_8
    got = 0_8
    full = ishft(1_8, N) - 1_8
    draws = 0_8
    do while (got /= full)
       k = draw(s, N)
       got = ior(got, ishft(1_8, k))
       draws = draws + 1_8
    end do
  end function one_trial
end module gacha_mod

program gacha
  use gacha_mod
  character(len=32) :: arg
  integer :: N, k
  integer(8) :: T, t_
  real(8) :: sum, sumsq, mean, var, H
  N = 10
  T = 1000000_8
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) N
  end if
  if (command_argument_count() >= 2) then
     call get_command_argument(2, arg); read (arg, *) T
  end if
  sum = 0.0d0; sumsq = 0.0d0

  ! T 回の試行は互いに独立. 各試行の引き回数を集計する.
  ! BEGIN ANSWER: 各試行は独立なので !$omp parallel do reduction(+:sum,sumsq) で並列化・集計せよ.
  !$omp parallel do reduction(+:sum,sumsq)
  ! END ANSWER
  do t_ = 1, T
     block
       integer(8) :: d
       d = one_trial(N, t_ * 2654435761_8 + 12345_8)
       sum = sum + real(d, 8)
       sumsq = sumsq + real(d, 8) * real(d, 8)
     end block
  end do
  ! BEGIN ANSWER: 上の parallel do を閉じる !$omp end parallel do を書け.
  !$omp end parallel do
  ! END ANSWER

  mean = sum / T
  var  = sumsq / T - mean * mean
  ! 理論期待値 = N * H_N
  H = 0.0d0
  do k = 1, N
     H = H + 1.0d0 / k
  end do
  print "(a,i0,a,i0,a,f0.3,a,f0.3,a,f0.3)", &
       "N=", N, ", trials=", T, ": 平均 ", mean, " 回 (理論 ", N * H, "), 標準偏差 ", sqrt(var)
end program gacha
